File: Python/Problem_105.py
def rule2_quick(a):
    a = sorted(a)
    n = len(a)
    # prefix sums for speed
    pref = [0]
    for x in a:
        pref.append(pref[-1] + x)

    for k in range(1, n // 2 + 1):
        left = pref[k + 1]              # a1..a(k+1)
        right = sum(a[-k:])             # largest k
        if left <= right:
            return False
    return True

def rule1_disjoint(a):
    n = len(a)
    seen = {}  # sum -> bitmask
    for mask in range(1, 1 << n):
        s = 0
        m = mask
        i = 0
        while m:
            if m & 1:
                s += a[i]
            i += 1
            m >>= 1

        prev = seen.get(s)
        if prev is None:
            seen[s] = mask
        else:
            if (prev & mask) == 0:  # disjoint subsets with equal sum
                return False
    return True

def is_special(a):
    a = sorted(a)
    return rule2_quick(a) and rule1_disjoint(a)

def solve(path="sets.txt"):
    ans = 0
    with open(path, "r", encoding="utf-8") as f:
        for line in f:
            a = list(map(int, line.strip().split(",")))
            if is_special(a):
                ans += sum(a)
    return ans

if __name__ == "__main__":
    print(solve("sets.txt"))  # 73702 

File: Java/Problem_105.java
import java.io.*;
import java.util.*;

public class Main {
    static boolean rule2Quick(int[] a) {
        int[] b = a.clone();
        Arrays.sort(b);
        int n = b.length;

        long[] pref = new long[n + 1];
        for (int i = 0; i < n; i++) pref[i + 1] = pref[i] + b[i];

        for (int k = 1; k <= n / 2; k++) {
            long left = pref[k + 1]; // smallest k+1
            long right = 0;
            for (int i = 0; i < k; i++) right += b[n - 1 - i]; // largest k
            if (left <= right) return false;
        }
        return true;
    }

    static boolean rule1Disjoint(int[] a) {
        int n = a.length;
        HashMap<Long, Integer> seen = new HashMap<>(1 << n);

        for (int mask = 1; mask < (1 << n); mask++) {
            long s = 0;
            int m = mask, i = 0;
            while (m != 0) {
                if ((m & 1) != 0) s += a[i];
                i++;
                m >>= 1;
            }

            Integer prev = seen.get(s);
            if (prev == null) {
                seen.put(s, mask);
            } else {
                if ((prev & mask) == 0) return false; // disjoint equal sum
            }
        }
        return true;
    }

    static boolean isSpecial(int[] a) {
        Arrays.sort(a);
        return rule2Quick(a) && rule1Disjoint(a);
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new FileReader("sets.txt"));
        String line;
        long ans = 0;

        while ((line = br.readLine()) != null) {
            String[] parts = line.trim().split(",");
            int[] a = new int[parts.length];
            for (int i = 0; i < parts.length; i++) a[i] = Integer.parseInt(parts[i]);

            if (isSpecial(a)) {
                long s = 0;
                for (int v : a) s += v;
                ans += s;
            }
        }
        br.close();

        System.out.println(ans); // 73702 
    }
}
